vllm/core/block/prefix_caching_block.py

```diff
@@ -688,7 +832,129 @@ def hash_block_tokens(is_first_block: bool, prev_block_hash: Optional[int],
         return hash((is_first_block, prev_block_hash, *cur_block_token_ids))
 
 
+class ComputedBlocksTracker:
+    """Handles caching of per-sequence computed block ids. 
+        When a sequence appears for the first time, it traverses all of the 
+        blocks and detects the prefix of blocks that is computed. On the
+        subsequent times, it only traverses the new blocks that were added 
+        and updates the already recorded prefix of blocks with the newly 
+        computed blocks.
+
+        To avoid redundant traversals, the algorithm also detects when there
+        is a "gap" in the computed prefix. For example, if we have blocks =
+        [1,2,3,4,5], and we have detected [1,2,3] as the computed prefix, then
+        we won't try to add more computed blocks to [1,2,3] in this sequence
+        iteration, and will add more computed blocks only after the sequence is
+        freed and reused again.
+
+        Note that currently, for a given sequence, we also skip the last 
+        block id for caching purposes, to avoid caching of a full sequence
```

Comment:
does this work with lookahead scheduling (where potenially >1 block is modified in single step)? don't have to fix now but in the future we want speculative decoding x prefix caching to work

Reply:
I think it should work since the blocks that are used for appending or speculative tokens won't be marked as computed, so they won't go into the common cache prefix.